src/java/com/puppetlabs/jruby_utils/pool/ReferencePool.java

```diff
@@ -0,0 +1,526 @@
+
+    @Override
+    public void register(E e) {
+        final ReentrantLock lock = this.borrowLock;
+        lock.lock();
+        try {
+            if (instance != null) {
+                throw new IllegalStateException(
+                        "Unable to register additional instance, pool full");
+            }
+
+            instance = e;
+            currentBorrowCount.set(0);
+
+            signalPoolNotEmpty();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Unregisters the JRuby instance. Blocks waiting for all borrows of the
+     * instance to be returned before clearing it out. In this pool implementation,
+     * `clear` and `unregister` are aliases of one another, since clearing the pool
+     * is the same as clearing the on active instance.
+     *
+     * @param e the instance to clean up
+     * @throws InterruptedException
+     */
+    @Override
+    public void unregister(E e) throws InterruptedException {
+        final ReentrantLock lock = this.borrowLock;
+        lock.lock();
+        try {
+            if (currentBorrowCount.get() != 0) {
+                instanceNotBorrowed.await();
+            }
+            instance = null;
+
+            signalIfLockCanProceed();
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public E borrowItem() throws InterruptedException {
+        E item = null;
+        final ReentrantLock lock = this.borrowLock;
+        lock.lock();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            do {
+                if (this.pill != null) {
+                    // Return the pill immediately if there is one
+                    item = pill;
+                } else if (isPoolLockHeldByAnotherThread(currentThread)) {
+                    poolNotLocked.await();
+                } else if (instance == null) {
+                    // No instance initialized yet
+                    borrowsAvailable.await();
+                } else if (this.currentBorrowCount.get() >= this.maxBorrowCount) {
+                    // Max borrow count reached, wait for one to be returned
+                    borrowsAvailable.await();
+                } else {
+                    item = this.instance;
+                    this.currentBorrowCount.getAndIncrement();
+                    LOGGER.info("Borrowing, count is " + currentBorrowCount.get());
+                }
+            } while (item == null);
+        } finally {
+            lock.unlock();
+        }
+
+        return item;
+    }
+
+    @Override
+    public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws
+            InterruptedException {
+        E item = null;
+        final ReentrantLock lock = this.borrowLock;
+        long remainingMaxTimeToWait = unit.toNanos(timeout);
+
+        // `borrowLock.lockInterruptibly()` is called here as opposed to just
+        // `borrowLock.borrowLock` to follow the pattern that the JDK's
```

Comment:
Should these be `lock.<blah>`?
And should the second one be `lock.lock()`? 
I'm looking at the methods above.

Reply:
I'm gonna change them to just reference the method name.